spec/upstream_spec.lua

```diff
@@ -0,0 +1,152 @@
+local Upstream = require('apicast.upstream')
+local match = require('luassert.match')
+
+describe('Upstream', function()
+    local valid_url = 'http://localhost:8080/path?query'
+
+    describe('.new', function()
+        it('returns an error on invalid upstream', function()
+            assert.returns_error('invalid upstream', Upstream.new('invalid uri'))
+        end)
+
+        it('returns new instance', function()
+            assert(Upstream.new(valid_url))
+        end)
+
+        it('returns table with location_name', function()
+            -- @upstream location is defined in the apicast.conf file
+            assert.equal('@upstream', Upstream.new(valid_url).location_name)
+        end)
+
+        it('returns table with upstream_name', function()
+            -- upstream name is defined in the upstream.conf file
+            assert.equal('upstream', Upstream.new(valid_url).upstream_name)
+        end)
+
+        it('returns table with resolver', function()
+            assert.equal(require('resty.resolver'), Upstream.new(valid_url).resolver)
+        end)
+    end)
+
+    describe(':resolve', function()
+        it('uses .resolver', function()
+            local upstream = Upstream.new(valid_url)
+            local resolver, instance, servers = {}, {}, {}
+            stub(resolver, 'instance').returns(instance)
+            stub(instance, 'get_servers').returns(servers)
+
+            upstream.resolver = resolver
+
+            assert.equal(servers, upstream:resolve())
```

Comment:
I'm not sure it's a good idea to test with servers = {}
If `upstream:resolve()` fails somehow and returns an empty table we won't notice.

Reply:
`.equal` is by reference, so it needs to be exactly the same table

but I'll add there a value to make that clearer